### PR TITLE
bpo-39622: Interrupt the main asyncio task on Ctrl+C

### PR DESCRIPTION
**Pull request by asvetlov**
The strategy is:
1. If asyncio runner is executed in the main thread it can handle Ctrl+C
2. Ctrl+C cancels the main task (created for executing `runner.run(coro())`).
3. The second and following Ctrl+C  don't cancel the main task but raise KeyboardInterrupt instead. It can break loops, blocking calls etc.
4.  `runner.run()` raises `KeyboardInterrupt` when the main task is cancelled by Ctrl+C
5. Background tasks are not touched, they are cancelled at the runner finalization stage.

https://bugs.python.org/issue39622

### CODE REVIEW

**Comment by asvetlov** [comment]
@1st1 @gvanrossum the PR adds Ctrl+C handling to asyncio programs.
Please review if you have time.

File: Lib/asyncio/runners.py
Comment (kumaraditya303) [inline Lib/asyncio/runners.py]:
```suggestion
            # wakeup loop if it is blocked by select() with long timeout
```

Comment (asvetlov) [reply Lib/asyncio/runners.py]: Thanks!

**Review by kumaraditya303 — APPROVED** [review]
LGTM, tested on linux.

File: Lib/asyncio/runners.py
Comment (gvanrossum) [inline Lib/asyncio/runners.py]: Is this a typo for `_interrupt_count`?

File: Doc/library/asyncio-runner.rst
Comment (asvetlov) [inline Doc/library/asyncio-runner.rst:124]: @kumaraditya303 could you review this section, please?
Feel free to suggest wild rephrasing if you don't like the chapter structure.

File: Doc/library/asyncio-runner.rst
Comment (kumaraditya303) [inline Doc/library/asyncio-runner.rst]:
```suggestion
1. :meth:`asyncio.Runner.run` installs a custom :const:`signal.SIGINT` handler before
   any user code is executed and removes it when exiting from the function.
```

File: Doc/library/asyncio-runner.rst
Comment (kumaraditya303) [inline Doc/library/asyncio-runner.rst]:
```suggestion
2. The :class:`~asyncio.Runner` creates the main task for the passed coroutine for 
    its execution.
```

File: Doc/library/asyncio-runner.rst
Comment (kumaraditya303) [inline Doc/library/asyncio-runner.rst]:
```suggestion
3. When :const:`signal.SIGINT` is raised by :kbd:`Ctrl-C`, the custom signal handler
   cancels the main task by calling :meth:`asyncio.Task.cancel` which raises 
   :exc:`asyncio.CancelledError` inside the the main task.  This causes the Python stack 
   to unwind,  ``try/except`` and ``try/finally`` blocks can be used for resource cleanup. 
   After the main task is cancelled, :meth:`asyncio.Runner.run` raises :exc:`KeyboardInterrupt`.
```

File: Doc/library/asyncio-runner.rst
Comment (kumaraditya303) [inline Doc/library/asyncio-runner.rst]:
```suggestion
4. A user could write a tight loop which cannot be interrupted by :meth:`asyncio.Task.cancel`,
   in which case the second following :kbd:`Ctrl-C` immediately raises the :exc:`KeyboardInterrupt`
   without cancelling the main task.
```

File: Doc/library/asyncio-runner.rst
Comment (kumaraditya303) [inline Doc/library/asyncio-runner.rst]:
```suggestion
When :const:`signal.SIGINT` is raised by :kbd:`Ctrl-C`, :exc:`KeyboardInterrupt`
```

File: Doc/library/asyncio-runner.rst
Comment (kumaraditya303) [inline Doc/library/asyncio-runner.rst]:
```suggestion
exception is raised in the main thread by default. However this doesn't work with :mod:`asyncio` 
because it can interrupt asyncio internals and can hang the program from exiting.

To mitigate this issue, :mod:`asyncio` handles :const:`signal.SIGINT` as follows:
```

File: Doc/library/asyncio-runner.rst
Comment (kumaraditya303) [inline Doc/library/asyncio-runner.rst]:
```suggestion

```

File: Doc/library/asyncio-runner.rst
Comment (kumaraditya303) [reply Doc/library/asyncio-runner.rst:124]: Sure, I have added inline suggestions.

File: Lib/asyncio/runners.py
Comment (kumaraditya303) [inline Lib/asyncio/runners.py]:
```suggestion
        self._interrupt_count = 0
```

File: Lib/asyncio/runners.py
Comment (kumaraditya303) [inline Lib/asyncio/runners.py]:
```suggestion
        self._interrupt_count = 0
```

File: Lib/asyncio/runners.py
Comment (kumaraditya303) [inline Lib/asyncio/runners.py]:
```suggestion
            if self._interrupt_count > 0 and task.uncancel() == 0:
```

File: Lib/asyncio/runners.py
Comment (kumaraditya303) [inline Lib/asyncio/runners.py]:
```suggestion
        self._interrupt_count += 1
```

File: Lib/asyncio/runners.py
Comment (kumaraditya303) [inline Lib/asyncio/runners.py]:
```suggestion
        if self._interrupt_count == 1 and not main_task.done():
```

**Comment by asvetlov** [comment]
Thank you very much, @kumaraditya303 !

**Review by gvanrossum — COMMENTED** [review]
I'm trying this with code like this:
```py
async def looper():
    while True:
        print("Top...")
        try:
            await asyncio.sleep(1.0)
        except BaseException as err:
            print("Caught", repr(err))
```
and its behavior is very unpredictable. The first time I hit ^C it will show that it caught CancelledError, maybe the second time too, but after that something weird happens: it shows the KeyboardInterrupt (in runners.py on L142) causing another KeyboardInterrupt (selectors.py L561 on macOS, somewhere in windows_events.py on Windows), then the code prints "Caught GeneratorExit()", and then an endless loop of alternating "Top..." and "Caught RuntimeError('no running event loop')" until I hit another ^C, which stops it.

If I special-case GeneratorExit to break out of the loop, there is no endless loop, but I see something like this (on Windows):
```
Traceback (most recent call last):
  File "C:\Users\gvanrossum\pr-32105\Lib\asyncio\runners.py", line 175, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "C:\Users\gvanrossum\pr-32105\Lib\asyncio\runners.py", line 109, in run
    return self._loop.run_until_complete(task)
    super().run_forever()
    ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gvanrossum\pr-32105\Lib\asyncio\base_events.py", line 604, in run_forever
    self._run_once()
    ^^^^^^^^^^^^^^^^
  File "C:\Users\gvanrossum\pr-32105\Lib\asyncio\base_events.py", line 1873, in _run_once
    event_list = self._selector.select(timeout)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gvanrossum\pr-32105\Lib\asyncio\windows_events.py", line 439, in select
    self._poll(timeout)
    ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\gvanrossum\pr-32105\Lib\asyncio\windows_events.py", line 808, in _poll
    status = _overlapped.GetQueuedCompletionStatus(self._iocp, ms)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt
Catching Generator Exit
Task was destroyed but it is pending!
task: <Task cancelling name='Task-1' coro=<looper() done, defined at C:\Users\gvanrossum\pr-32105\t.py:3> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[gather.<locals>._done_callback() at C:\Users\gvanrossum\pr-32105\Lib\asyncio\tasks.py:759]>
^CTerminate batch job (Y/N)? y
PS C:\Users\gvanrossum\pr-32105>
```
This does not seem to be a big improvement over the original issue. :-(

**Comment by 1st1** [comment]
Why is this merged? No core dev reviewed this. Please revert.

**Comment by davidhewitt** [comment]
This appears like it might have caused breakage in the latest alpha: https://github.com/PyO3/pyo3/pull/2314#issue-1207426789

**Comment by gvanrossum** [comment]
> This appears like it might have caused breakage in the latest alpha: [PyO3/pyo3#2314 (comment)](https://github.com/PyO3/pyo3/pull/2314#issue-1207426789)

I find the discussion in that issue hard to follow. Sure, something that happened between a6 and a7 broke PyO3. But what? Why would this change (which only affects how KeyboardInterrupt is handled) be the root cause?

**Comment by davidhewitt** [comment]
@gvanrossum sorry for my short comment earlier; was up in the middle of the night helping my wife feed my newborn :)

I haven't had chance to repro and poke around with this this myself, however I think the cause is possibly because the `signal` and `threading` modules are disagreeing about what is the "main thread" in PyO3's unit test suite. (Rust binary embedding Python via dynamic linking to `libpython`.)

The line erroring is the `runners.py` line 103 introduced in this patch, which installs the sigint handler: https://github.com/python/cpython/blob/2e49bd06c5ffab7d154042da4b5bcd5ae6e4220b/Lib/asyncio/runners.py#L103

It reaches this error in the C: https://github.com/python/cpython/blob/b6d5e3c3c92374598b327918c8e66e748a1a92a0/Modules/signalmodule.c#L533-L535

That `runners.py` line 103 is guarded by a check `threading.current_thread() is threading.main_thread()`, hence my guess that `threading` / `signal` disagree what the main thread is in this case.

---

The PyO3 unit test suite tickling this crash can roughly be summarised as:
 - Rust binary embedding Python (dynamic linking to `libpython`)
 - The Python interpreter is started with `Py_InitializeEx(0)` (`0` means don't register signal handlers)
 - A _different_ native thread acquires the GIL with `PyGILState_Ensure()`
 - That thread then uses `PyEval_EvalCode` to run a snippet which contains a call to `asyncio.run`

**Comment by gvanrossum** [comment]
Sorry, it looks like you just answered at least part of my question on the PyO3 issue. Anyway, it looks like we need a better check for determining whether it is safe to call signal(). Or perhaps we just need to put a try/except around that call?

**Comment by davidhewitt** [comment]
Thinking on this further, I suspect that `threading` check passes because running this code snippet is probably the first time `threading` is imported, which would explain why `threading` module thinks the thread is the main thread.

I see in `signalmodule.c` there's a symbol `_PyOS_IsMainThread`, which looks like the correct check (used in C) for whether `signal.signal` can be called. Do you think it makes sense to expose this as a Python API, perhaps `signal.can_handle_signals()`?

**Comment by gvanrossum** [comment]
@davidhewitt We're 1-2 weeks from feature freeze, and I don't have time to propose and implement a new API for this purpose (simple though it may seem, there are always loose ends, tests to write, etc.).

What I *can* do (other than rolling back @asvetlov's PR, which would be a shame) is add a try/except around the `signal()` call that catches the ValueError. That would presumably unblock you. (Things would go even faster if you proposed a PR -- I can just merge it.)

**Comment by gvanrossum** [comment]
(Either way you should probably open a new issue about this rather than keep chatting on this closed PR.)

**Comment by davidhewitt** [comment]
👍 I'll do my best to find time to write a fresh issue and PR at the weekend. Thanks for the help!

**Pull request merged**
